codecov_auth/views/okta.py

```diff
@@ -0,0 +1,191 @@
+from urllib.parse import urlencode
+
+import jwt
+import requests
+from django.conf import settings
+from django.contrib.auth import login, logout
+from django.http import HttpRequest, HttpResponse
+from django.shortcuts import redirect
+from django.views import View
+from requests.auth import HTTPBasicAuth
+
+from codecov_auth.models import OktaUser, User
+from codecov_auth.views.base import LoginMixin
+from utils.services import get_short_service_name
+
+log = logging.getLogger(__name__)
+
+
+def validate_id_token(iss: str, id_token: str) -> dict:
+    res = requests.get(f"{iss}/oauth2/v1/keys")
+    jwks = res.json()
+
+    public_keys = {}
+    for jwk in jwks["keys"]:
+        kid = jwk["kid"]
+        public_keys[kid] = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk))
+
+    print(jwt.get_unverified_header(id_token))
+
+    kid = jwt.get_unverified_header(id_token)["kid"]
+    key = public_keys[kid]
+
+    id_payload = jwt.decode(
+        id_token,
+        key=key,
+        algorithms=["RS256"],
+        audience=settings.OKTA_OAUTH_CLIENT_ID,
+    )
+    assert id_payload["iss"] == iss
+    assert id_payload["aud"] == settings.OKTA_OAUTH_CLIENT_ID
+
+    return id_payload
+
+
+auth = HTTPBasicAuth(settings.OKTA_OAUTH_CLIENT_ID, settings.OKTA_OAUTH_CLIENT_SECRET)
+
+
+class OktaLoginView(LoginMixin, View):
+    def _fetch_user_data(self, iss: str, code: str) -> Optional[Dict]:
+
+        res = requests.post(
+            f"{iss}/oauth2/v1/token",
+            auth=auth,
+            data={
+                "grant_type": "authorization_code",
+                "code": code,
+                "redirect_uri": settings.OKTA_OAUTH_REDIRECT_URL,
+            },
+        )
+
+        if res.status_code >= 400:
+            return None
+        return res.json()
+
+    def _redirect_to_consent(self, iss: str) -> HttpResponse:
+        qs = urlencode(
+            dict(
+                response_type="code",
+                client_id=settings.OKTA_OAUTH_CLIENT_ID,
+                scope="openid email profile",
+                redirect_uri=settings.OKTA_OAUTH_REDIRECT_URL,
+                state=iss,
+            )
+        )
+        redirect_url = f"{iss}/oauth2/v1/authorize?{qs}"
+        response = redirect(redirect_url)
+        self.store_to_cookie_utm_tags(response)
+
+        if settings.OKTA_ISS is None:
+            # ISS was passed to us from Okta
+            self._store_iss_cookie(iss, response)
+
+        return response
+
+    def _store_iss_cookie(self, iss: str, response: HttpResponse):
+        response.set_cookie(
+            "_okta_iss",
+            iss,
+            max_age=86400,  # Same as state validatiy
+            httponly=True,
+            domain=settings.COOKIES_DOMAIN,
+        )
+
+    def _perform_login(self, request: HttpRequest) -> HttpResponse:
+        code = request.GET.get("code")
+        iss = settings.OKTA_ISS or request.COOKIES.get("_okta_iss")
+        if iss is None:
```

Comment:
Could we ever hit this? Since at the very least we'd have the iss from `_redirect_to_consent`

Reply:
I'd say this would be unexpected.  IDK - maybe somebody has cookies disabled or something.

Reply:
Yeaah that's fair, asking in case this could be dead code that adds complexity, but I think that's fair. In that scenario they wouldn't be able to use okta hmmm, unlikely, but we can keep it in here and see if we hit this